feat(rating): Add rating component

macandcheese opened the pull request:
Adds scaffold for rating component and initial functionality

**Related Issue:** Resolves https://github.com/Esri/calcite-components/issues/818

## Summary
New rating component!
![Screencast 2020-08-05 at 4 31 09 PM](https://user-images.githubusercontent.com/4733155/89474069-67a20300-d739-11ea-8620-c51b4afce88c.gif)

To-do:
- [x] Emit events for rating change
- [x] Add tests
- [x] Implement partial average display
- [x] Clean up styles and display of value - not totally happy with that
- [x] Any other style things cc @bstifle finds
- [x] Open to API 💭 or critiques

<!--

Please make sure the PR title and/or commit message adheres to the https://conventionalcommits.org/ specification.

Note: If your PR only has one commit and it is NOT semantic, you will need to either

a. add another commit and wait for the check to update
b. proceed to squash merge, but make sure the commit message is the same as the title.

This is because of the way GitHub handles single-commit squash merges (see https://github.com/zeke/semantic-pull-requests/issues/17)

If this is component-related, please verify that:

- [ ] feature or fix has a corresponding test
- [ ] changes have been tested with demo page in Edge

-->

paulcpederson commented:
Nice! I think there is a case where we show an item's rating but you haven't rated yet, and then the rating you gave it. Not sure if this design has a way to distinguish between those ideas?

macandcheese commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,216 @@
+  @Prop({ reflect: true }) theme: "light" | "dark";
+
+  /** specify the icon used for the rating, defaults to star */
+  @Prop({ reflect: true }) iconType: "star" | "circle" = "star";
```
Right now it only works with the icons that have reciprocal "-f" versions... For now circle and star. 

In places where we lock down icon selections to a few (accordion, split button), 'iconType' has been used to suggest that not any icon should used. Also why it has to be validated since I'm using the filled version of icons.

Could certainly look at allowing any icons and just styling with color but since all the icons are outlined now it's not really clear what's selected 😟.

macandcheese replied:
Kind of indifferent on that ... not sure there will be more added anytime soon since they just nixed basically all the filled icons 😆, although if someone does have a need for banana ratings they would probably be able to create a `banana-f` icon... 

These two seem reasonable for rating UI for now but not sure why we wouldn't document the two choices 🤷‍♂️ ...

macandcheese replied:
We could also just remove this choice and limit to only use stars for now until someone asks for circles or another icon explicitly...

macandcheese commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,216 @@
+  @Prop({ reflect: true }) value = 0;
+
+  /** the precision of the rating component - half or whole step */
+  @Prop({ reflect: true }) precision: "half" | "whole" = "whole";
```
Yeah that's a good improvement. I had thought about this as letting users select half steps but maybe that's a separate config option. To your point it should support display of any value.

macandcheese replied:
Just more thoughts here, this was initially added to allow users to select half - star ratings... I don't think there's a use case for users choosing 2.25 or 2.75 but 2.5 is reasonable (maybe?)

However I'm kind of inclined to just always show the exact value provided if not user entered.... Ie if an average rating is 4.21 should 21% of the final icon be filled or do we want to provide an option to round that in some way visually?

driskull replied:
> re thoughts here, this was initially added to allow users to select half - star ratings... I don't think there's a use case for users choosing 2.25 or 2.75 but 2.5 is reasonable (maybe?)

I thought the decimal was for rounding the rating from all users not for a single user to rate a half star? Do we really need to let users give half stars?

macandcheese replied:
I've removed this prop in latest update... Dev-users can just pass in the average they want, the stars will always display the partial average as requested. Can look at again later if needed.

macandcheese commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,216 @@
+  @Prop({ reflect: true }) precision: "half" | "whole" = "whole";
+
+  /** is the rating component in a selectable mode */
+  @Prop({ reflect: true }) readOnly = false;
```
Use case for read only would be... A user has just submitted a rating and that shouldn't appear transparent like a disabled one. But, yeah they are similar enough only the one is needed and it's confusing to have both, ha.

macandcheese commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,216 @@
+  connectedCallback() {
+    // prop validations
+    // ensure only allowed icons are used
+    const icon = ["star", "circle"];
```
See above about needing the reciprocal "-f" versions of icons. But yeah I think being more flexible would be a good thing here if there's a way.

macandcheese commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  componentDidLoad() {
+    this.ratingItems = this.el.shadowRoot.querySelectorAll("calcite-icon");
```
Yeah I think I should update that, might tie to your above feedback too 🤙

macandcheese commented:
> Nice! I think there is a case where we show an item's rating but you haven't rated yet, and then the rating you gave it. Not sure if this design has a way to distinguish between those ideas?

Hm, any screenshots? Didn't consider it but we should handle that somehow. Does it display both the previous rating and the user-submitted one? Or just the updated value once the user submits one?

paulcpederson commented:
OK, so I went and looked at how ratings are done in AGO, seems like they always display the actual rating (ie everybody's rating) but on hover shift the color to yellow and show your rating? IDK, open to ideas. Essentially the feature is a way to display "everybody" vs "me".

<img width="141" alt="Screen Shot 2020-08-06 at 10 30 28 AM" src="https://user-images.githubusercontent.com/1031758/89563410-42ac9f00-d7d0-11ea-9699-c9dbed6c5267.png">
<img width="132" alt="Screen Shot 2020-08-06 at 10 30 21 AM" src="https://user-images.githubusercontent.com/1031758/89563413-43ddcc00-d7d0-11ea-9ac9-88cf6870e64d.png">

paulcpederson commented:
Another interesting consideration, when I see a number, that's the "everybody" rating, right?

macandcheese commented:
Currently the number just shows the user-entered value. But, maybe that should be configurable, or not handled by the component?

That's often a place where a rating count would be shown too... (122 ratings) or something.

We could have an initial-value mode that's overrideable by a user (maybe with Grey-filled icons) but I think once a user chooses their own value that's the only thing that should display?

driskull commented:
This is how I would expect it to work:

- Initially displays cumulative rating
- User hovers/focuses or clicks an edit button to activate the "enter rating" functionality and user rating is shown
- Once user has submitted rating/blurs/mouseout then the cumulative rating is displayed again.

Does that sound correct?

Maybe there should be a prop for the total number of ratings to display that?

macandcheese commented:
I'm inclined to just remove the `display-value` prop and replace it with a prop to pass a rating count. 

I don't think showing the number of stars you have selected.... next to the number of stars you have selected does much for a user. I think that a tooltip could be shown on hover that had the exact value if that's needed at all.

There may be an elegant way to handle both for a case where a dev can provide both? 
![Slice](https://user-images.githubusercontent.com/4733155/89566159-81445880-d7d4-11ea-8018-bee612e6fa1a.png)

macandcheese commented:
OK API update... and some more ❓ s....

- 🗑️ - Removed `display-rating` prop

- ➕ - Added `rating-average` and `rating-count`. If either are present, the chip will display.
  - I don't think we want to handle any calculation on submission here, we should just rely on consumers using change event to update the `rating-average` and `rating-count` props themselves

- ❓ - I think the `count` prop that determines number of icons could be changed now, it's a little confusing with `rating-count`. `length` ? `total` ?
  - Do we want to limit the `count` prop in any way?

- ❓ - There is now no display of user-entered value. The big blue filled icons should be enough to differentiate between the `rating-average` and value, I think this is fine?

- ❓ - Currently `value` can be present in DOM on load, but gets replaced when user submits. I think we could rely on the `rating-average` to populate an initial value, and then, to Paul's point, style it differently until a user submits there own `value`, which would take display precedence.

![Screencast 2020-08-06 at 11 53 58 AM](https://user-images.githubusercontent.com/4733155/89571441-860d0a80-d7dc-11ea-9b37-bdec0c45c8d1.gif)

driskull commented:
>  I think the count prop that determines number of icons could be changed now, it's a little confusing with rating-count. length ? total ?

I think `ratings` or `total` sound good.

> There is now no display of user-entered value. The big blue filled icons should be enough to differentiate between the rating-average and value, I think this is fine?

Sounds good to me :). maybe `average` is fine instead of `rating-average`?

> Currently value can be present in DOM on load, but gets replaced when user submits. I think we could rely on the rating-average to populate an initial value, and then, to Paul's point, style it differently until a user submits there own value, which would take display precedence.

When a user submits it would throw an event and the developer could use that to update the total and average, right?

macandcheese commented:
Updates: 
- number of icon items is now determined by `total` prop
- renamed some props : `average`, `count` (number of ratings to display in chip, just a display prop)
- updated how the `average` and `value` props play together to allow for use cases like @paulcpederson mentioned, it works as such:


  - If `calcite-rating` is `read-only` or `disabled` : `average` and `value` will both display as blue, with the `value` prop taking display precedence because it represents a user-entered value.
  - Otherwise, `value` will always be shown in blue, if present. If `value` is not present, but `average` is, the `average` will be shown in yellow until a user interacts and sets their own `value`, which is preserved in the UI and a `calciteRatingChange` event allows devs to use that rating to update count and average.
 - It's a little confusing but I think it allows for a good amount of flexibility for use cases.

- I updated dev demo with examples but a visual is more helpful:
![Screencast 2020-08-07 at 9 53 00 AM](https://user-images.githubusercontent.com/4733155/89669222-ea3ed580-d893-11ea-8835-c87d96db3d52.gif)
 


Still to do
- tests
- partial display of ratings for `average` (I think I agree with @driskull that users should just select whole steps, for now at least)
- clean up

macandcheese commented:
Update: `read-only` now shows filled stars to better indicate it's in a read-only state:
<img width="899" alt="Screen Shot 2020-08-07 at 10 36 15 AM" src="https://user-images.githubusercontent.com/4733155/89672672-e910a700-d899-11ea-9c03-7f3452e3ffd5.png">

driskull commented:
Lookin pretty nice @macandcheese 🎉

bstifle commented:
Anyone have objections to limiting this to 5 and 10 stars/circles only?

If so, please point me to an active use case where a different number for rating is used, thanks!

bstifle commented:
OR, do we even need 10? all in favor for only using 5 stars/circles only?

driskull commented:
I would just leave it hard coded at 5 for now. We can always add a prop later which is better than breaking a change later.

The only reason for having 10 would be if someone wanted the accuracy of 5 stars + half stars. :)

bstifle commented:
perfect. 5 it is.

macandcheese commented:
Roger that. I'll update to remove that option and set everything at 5 in my next pass.

macandcheese commented:
Latest commit handles partial ratings and overriding them (rather crudely... for now), handles wrapping labels, and limits to 5 stars.

macandcheese commented:
Tests added.. Need to solve some style issues with clip-path in Safari, fix Edge bugs, make ie11 work ( 😦 ).

Open to better ideas for the generation and placement of the partial rating star...

macandcheese commented:
Apologies for letting this linger:

This line is giving me issues:
https://github.com/Esri/calcite-components/pull/821/files#diff-a783f88f49046472f395677e01b66a7fR245

Typescript is saying it is not a valid CSS declaration but it works in practice and seems to be the only way to set clipPath for Safari...

I'm using clipPath to clip the exact % of the partially filled star for averages. I am also all ears to a way to do this in IE / Edge :)

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,356 @@
+
+      /*
+      // @ts-ignore-error
+      // this.partialStar.style.webkitClipPath = this.dir !== "rtl" ? ltrClip : rtlClip;
```
I think it would probably be ok to do this:

```ts
// set webkit vendor specific prefix
(this.partialStar.style as any).webkitClipPath = this.dir !== "rtl" ? ltrClip : rtlClip;
```

macandcheese replied:
Updated to just use width ⬅️ ➡️

macandcheese commented:
> @macandcheese can you use `overflow:hidden` and set a width to the percentage shown?

Yeah, simpler ended up being way better here. Updated that, and tested in Edge and IE11. Working and ready for re-review.

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,385 @@
+          onMouseEnter={(e) => this.showSelectedIconOnHover(e)}
+          onTouchStart={(e) => this.showSelectedIconOnHover(e)}
+          onFocus={(e) => this.showSelectedIconOnHover(e)}
+          tabindex={!this.readOnly && !this.disabled ? 0 : null}
```
Instead of querying for all the icons when the component is loaded, can you store the reference like so?

```tsx
ref={(el) => this.ratingItems[index] = el as HTMLCalciteIconElement}
```

driskull replied:
Maybe `ratingItems` could be a stateful prop incase the amount is configurable in the future?

macandcheese replied:
Yeah, good updates.

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,385 @@
+
+    // ie11 safe version of array / spread / keys
+    const iconsToRender = [];
+    for (let i = 0; i < 5; i++) {
```
nitpick: maybe make the `5` a constant in the top of the file?

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,385 @@
+  private determineRatingPosition(e) {
+    if (!this.readOnly) {
+      // handle edge and ie not interpreting mapped stars
+      const isEdgeOrIE =
```
nitpick: this kinda makes me squeamish. Is there a way to use feature detection instead of browser detection? I'm worried this could be problematic in the future.

driskull replied:
Maybe instead of storing the value on the icon, there could be an interface for the array of icons that handles that?

```
ratingItems: {icon: HTMLCalciteIconElement, value: number,...}
```

driskull replied:
I guess the other solution would be to create a `calcite-rating-item` that used `calcite-icon` internally and added props for value, hovered, average, selected?

driskull replied:
Or a functional component would maybe be best since it wouldn't be public.

macandcheese replied:
Re: `isEdgeOrIE` - Luckily I can strip all that out - Legacy Edge was struggling to interpret the type of element rendered through the rendered map of stars, that's not there anymore since IE can't handle it at all so I can remove these references.

macandcheese replied:
Re: value - you mean instead of `data-value`? Yeah, that makes sense to me

driskull replied:
Yeah, I was just thinking of storing them either in a typed object or functional component instead of using data attributes.

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,385 @@
+  private showSelectedIconOnHover(e) {
+    if (!this.readOnly) {
+      // handle edge and ie not interpreting mapped stars
+      const isEdgeOrIE =
```
nitpick: Seems like this could be a method since it's used twice.

macandcheese replied:
Done.

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,385 @@
+
+  private selectedIconType = "star-f";
+
+  private dir;
```
nitpick: can this have a type? `"ltr" | "rtl"`?

macandcheese commented:
@driskull having a bit of trouble combining indexed elements, refs, with the new typed object. If I can figure that out this should be good to go 😓

driskull commented:
Yeah that might just be too complicated. Can we just use a `Map` to hold the `calcite-icon`s?
 
```tsx
@State()
 ratingItems: Map<number, HTMLCalciteIconElement> = new Map();
```

Something like that? The number could be the value and the map could be cleared before creating new rating items.
 
The ref for each item would add to the map with the value assigned.

```tsx
ref={(el) => (this.ratingItems.set(value, el))}
```

driskull commented:
I'm not sure if clearing the map would trigger a stateful change though. So maybe replacing the map after clearing might be better.

macandcheese commented:
@driskull - made some changes based on your suggestion, can you take a look? Would like to clean up these merge conflicts and get this in.  ⭐ ⭐ ⭐ ⭐ 🌟

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,366 @@
+  @Prop({ reflect: true }) displayValue = false;
+
+  /** optionally pass a number of previous ratings to display */
+  @Prop({ reflect: true }) count?: number;
```
@macandcheese should `count` be watched to call `determineInitialRating()`?

macandcheese replied:
This isn't actually used to compute a rating, it's just passed to the display chip if requested... we wouldn't know the value of all of those ratings to actually calculate ourselves.

driskull replied:
Gotcha. looks good

driskull commented on src/components/calcite-rating/calcite-rating.tsx:
```diff
@@ -0,0 +1,366 @@
+  @Prop({ reflect: true }) value = 0;
+
+  /** is the rating component in a selectable mode */
+  @Prop({ reflect: true }) readOnly = false;
```
Are we keeping both readOnly and disabled?

macandcheese replied:
I feel like they are separate use cases?

Read-only: Useful if you want to show that a rating has been entered and perhaps not able to be changed by a user (submitting a review on ecommerce)

Disabled: Useful if the rating isn't available at this time to a user. Perhaps if the rating module needs to be disabled while a user is editing another portion of an item or isn't available yet.

https://bootstrap-vue.org/docs/components/form-rating#disabled
https://bootstrap-vue.org/docs/components/form-rating#readonly

The display is also a little different - readonly doesn't show the "could be filled affordance".

driskull reviewed (approved):
Looks good @macandcheese!

macandcheese commented:
Merging if no objections.
